File: platforms/android/arfitkit/src/main/kotlin/com/arfitkit/ARFitKit.kt
/**
 * ARFitKit - Android SDK for AR Virtual Try-On
 *
 * オープンソースのAR試着SDK
 * KotlinでネイティブなARCore統合を提供
 */

package com.arfitkit

import android.content.Context
import android.graphics.Bitmap
import android.view.SurfaceView
import androidx.annotation.MainThread
import androidx.lifecycle.LifecycleOwner
import com.google.ar.core.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*

/**
 * 衣服の種類
 */
enum class GarmentType {
    UNKNOWN,
    TSHIRT,
    SHIRT,
    JACKET,
    COAT,
    DRESS,
    PANTS,
    SHORTS,
    SKIRT
}

/**
 * エラーコード
 */
sealed class ARFitKitError(message: String) : Exception(message) {
    object InitializationFailed : ARFitKitError("Initialization failed")
    object CameraAccessDenied : ARFitKitError("Camera access denied")
    object GpuNotAvailable : ARFitKitError("GPU not available")
    object ModelLoadFailed : ARFitKitError("Failed to load model")
    object GarmentConversionFailed : ARFitKitError("Garment conversion failed")
    object InvalidImage : ARFitKitError("Invalid image")
    object SessionNotStarted : ARFitKitError("Session not started")
    object NetworkError : ARFitKitError("Network error")
    object ARCoreNotAvailable : ARFitKitError("ARCore not available")
}

/**
 * セッション設定
 */
data class SessionConfig(
    val targetFPS: Int = 60,
    val enableClothSimulation: Boolean = true,
    val enableShadows: Boolean = true,
    val maxGarments: Int = 3,
    val serverEndpoint: String = "",
    val useHybridProcessing: Boolean = true
)

/**
 * 衣服データ
 */
data class Garment(
    val id: String = java.util.UUID.randomUUID().toString(),
    val type: GarmentType,
    val image: Bitmap,
    var isLoading: Boolean = false,
    var conversionProgress: Float = 0f,
    internal var nativeGarment: Long = 0
)

/**
 * ボディポーズデータ
 */
data class BodyPose(
    val landmarks: List<FloatArray>,
    val visibility: List<Float>,
    val confidence: Float,
    val timestamp: Long
)

/**
 * ARFitKit メインクラス
 * AR試着機能を提供するプライマリインターフェース
 */
class ARFitKit(private val context: Context) {
    
    // State flows
    private val _isSessionActive = MutableStateFlow(false)
    val isSessionActive: StateFlow<Boolean> = _isSessionActive.asStateFlow()
    
    private val _currentFPS = MutableStateFlow(0f)
    val currentFPS: StateFlow<Float> = _currentFPS.asStateFlow()
    
    private val _currentPose = MutableStateFlow<BodyPose?>(null)
    val currentPose: StateFlow<BodyPose?> = _currentPose.asStateFlow()
    
    private val _activeGarments = MutableStateFlow<List<Garment>>(emptyList())
    val activeGarments: StateFlow<List<Garment>> = _activeGarments.asStateFlow()
    
    // Private properties
    private var config: SessionConfig = SessionConfig()
    private var arSession: Session? = null
    private var surfaceView: SurfaceView? = null
    
    private val processingScope = CoroutineScope(Dispatchers.Default + SupervisorJob())
    
    // Callbacks
    var onFrameProcessed: ((Bitmap) -> Unit)? = null
    var onPoseUpdated: ((BodyPose) -> Unit)? = null
    var onError: ((ARFitKitError) -> Unit)? = null
    
    // MARK: - Native Methods
    
    private external fun nativeInitialize(targetFPS: Int, enableClothSimulation: Boolean)
    private external fun nativeStartSession()
    private external fun nativeStopSession()
    private external fun nativeGetCurrentFPS(): Float
    private external fun nativeProcessFrame(bitmap: Bitmap, timestampNs: Long)
    private external fun nativeLoadGarment(bitmap: Bitmap, type: Int): String?
    private external fun nativeTryOn(garmentId: String)
    private external fun nativeRemoveAllGarments()

    companion object {
        init {
            try {
                System.loadLibrary("arfit_jni")
                android.util.Log.d("ARFitKit", "Native library loaded successfully")
            } catch (e: UnsatisfiedLinkError) {
                android.util.Log.e("ARFitKit", "Failed to load native library: ${e.message}")
            }
        }
    }
    
    /**
     * SDKを初期化
     * @param config セッション設定
     */
    fun initialize(config: SessionConfig = SessionConfig()) {
        this.config = config
        
        // Native initialize
        nativeInitialize(config.targetFPS, config.enableClothSimulation)
        
        android.util.Log.d("ARFitKit", "Initialized with config: $config")
    }
    
    /**
     * ARセッションを開始
     * @param surfaceView AR表示用のSurfaceView
     * @param lifecycleOwner ライフサイクルオーナー
     */
    @MainThread
    suspend fun startSession(surfaceView: SurfaceView, lifecycleOwner: LifecycleOwner) {
        if (_isSessionActive.value) return
        
        // Check ARCore availability
        val availability = ArCoreApk.getInstance().checkAvailability(context)
        if (availability == ArCoreApk.Availability.UNSUPPORTED_DEVICE_NOT_CAPABLE) {
            throw ARFitKitError.ARCoreNotAvailable
        }
        
        // Request ARCore installation if needed
        val installStatus = ArCoreApk.getInstance().requestInstall(
            context as android.app.Activity?,
            true
        )
        
        if (installStatus == ArCoreApk.InstallStatus.INSTALL_REQUESTED) {
            return // Will resume after install
        }
        
        // Create AR session
        arSession = Session(context).apply {
            val arConfig = Config(this).apply {
                updateMode = Config.UpdateMode.LATEST_CAMERA_IMAGE
                focusMode = Config.FocusMode.AUTO
            }
            configure(arConfig)
        }
        
        this.surfaceView = surfaceView
        _isSessionActive.value = true
        
        // Start native session
        nativeStartSession()
        
        android.util.Log.d("ARFitKit", "Session started")
        
        // Start processing loop
        processingScope.launch {
            processingLoop()
        }
    }
    
    /**
     * ARセッションを停止
     */
    fun stopSession() {
        nativeStopSession()
        arSession?.pause()
        arSession?.close()
        arSession = null
        surfaceView = null
        _isSessionActive.value = false
        _activeGarments.value = emptyList()
        nativeRemoveAllGarments()
        
        android.util.Log.d("ARFitKit", "Session stopped")
    }
    
    /**
     * 衣服を読み込む
     * @param bitmap 衣服の画像
     * @param type 衣服の種類（省略時は自動検出）
     * @return 読み込まれた衣服
     */
    suspend fun loadGarment(bitmap: Bitmap, type: GarmentType = GarmentType.UNKNOWN): Garment = withContext(Dispatchers.IO) {
        val garment = Garment(type = type, image = bitmap, isLoading = true)
        
        val id = nativeLoadGarment(bitmap, type.ordinal)
        if (id == null) {
             throw ARFitKitError.GarmentConversionFailed
        }
        
        val loadedGarment = garment.copy(id = id, isLoading = false, conversionProgress = 1f)
        return@withContext loadedGarment
    }
    
    /**
     * 衣服を試着
     * @param garment 試着する衣服
     */
    fun tryOn(garment: Garment) {
        if (!_isSessionActive.value) {
            throw ARFitKitError.SessionNotStarted
        }
        
        val currentList = _activeGarments.value.toMutableList()
        
        // Check max garments
        if (currentList.size >= config.maxGarments) {
            currentList.removeAt(0)
        }
        
        currentList.add(garment)
        _activeGarments.value = currentList
        
        nativeTryOn(garment.id)
        
        android.util.Log.d("ARFitKit", "Trying on garment: ${garment.type}")
    }
    
    /**
     * 衣服を削除
     * @param garment 削除する衣服
     */
    fun removeGarment(garment: Garment) {
        _activeGarments.value = _activeGarments.value.filter { it.id != garment.id }
        nativeRemoveGarment(garment.id)
    }
    
    /**
     * すべての衣服を削除
     */
    fun removeAllGarments() {
        _activeGarments.value = emptyList()
        nativeRemoveAllGarments()
    }
    
    /**
     * リソースを解放
     */
    fun release() {
        stopSession()
        processingScope.cancel()
    }
    
    /**
     * ARCore フレーム取得・処理ループ
     * セッションがアクティブな間、カメラフレームを取得し
     * Core C++ エンジンで処理（物理・レンダリング）した結果をコールバックに渡す
     */
    private suspend fun processingLoop() {
        // 処理結果の転送用Bitmap（毎フレーム再生成を避けるためキャッシュ）
        var outputBitmap: Bitmap? = null
        
        while (isActive && _isSessionActive.value) {
            try {
                val session = arSession ?: break
                
                // ARCoreから最新フレームを取得
                val frame = session.update()
                val camera = frame.camera
                
                // カメラのトラッキング状態を確認
                if (camera.trackingState != TrackingState.TRACKING) {
                    delay(16)
                    continue
                }
                
                // カメラ画像を取得してBitmapに変換
                val image = frame.acquireCameraImage()
                val width = image.width
                val height = image.height
                
                // 出力Bitmapのサイズが異なる場合は再作成
                if (outputBitmap == null || outputBitmap.width != width || outputBitmap.height != height) {
                    outputBitmap?.recycle()
                    outputBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
                }
                
                // YUV → RGBA変換してBitmapにコピー
                val yBuffer = image.planes[0].buffer
                val uBuffer = image.planes[1].buffer
                val vBuffer = image.planes[2].buffer
                
                val ySize = yBuffer.remaining()
                val uSize = uBuffer.remaining()
                val vSize = vBuffer.remaining()
                
                val nv21 = ByteArray(ySize + uSize + vSize)
                yBuffer.get(nv21, 0, ySize)
                vBuffer.get(nv21, ySize, vSize)
                uBuffer.get(nv21, ySize + vSize, uSize)
                
                val yuvImage = android.graphics.YuvImage(
                    nv21, android.graphics.ImageFormat.NV21, width, height, null
                )
                val out = java.io.ByteArrayOutputStream()
                yuvImage.compressToJpeg(android.graphics.Rect(0, 0, width, height), 90, out)
                val jpegBytes = out.toByteArray()
                val decodedBitmap = android.graphics.BitmapFactory.decodeByteArray(jpegBytes, 0, jpegBytes.size)
                
                // decodedBitmapをoutputBitmapにコピー
                val canvas = android.graphics.Canvas(outputBitmap!!)
                canvas.drawBitmap(decodedBitmap, 0f, 0f, null)
                decodedBitmap.recycle()
                
                image.close()
                
                // JNI経由でCore C++エンジンにフレームを処理させる
                // outputBitmapは in-place で書き換えられる
                nativeProcessFrame(outputBitmap, frame.timestamp)
                
                // FPSの更新
                val fps = nativeGetCurrentFPS()
                _currentFPS.value = fps
                
                // 処理済みフレームをコールバック経由でUIに配信
                withContext(Dispatchers.Main) {
                    onFrameProcessed?.invoke(outputBitmap!!)
                }
                
            } catch (e: com.google.ar.core.exceptions.NotYetAvailableException) {
                // フレームがまだ利用不可（ARCore初期化中）
                delay(16)
            } catch (e: Exception) {
                android.util.Log.e("ARFitKit", "Processing error: ${e.message}")
                delay(16)
            }
        }
        
        // ループ終了時にBitmapを解放
        outputBitmap?.recycle()
    }
    
    // ネイティブメソッド: 個別の衣服を削除
    private external fun nativeRemoveGarment(garmentId: String)
}

File: platforms/ios/Sources/ARFitKit/ARFitKit.swift
/**
 * ARFitKit - iOS SDK for AR Virtual Try-On
 *
 * オープンソースのAR試着SDK
 * SwiftでネイティブなARKit統合を提供
 */

import Foundation
import ARKit
import Combine

// MARK: - Public Types

/// 衣服の種類
public enum GarmentType: String, Codable, CaseIterable {
    case unknown
    case tshirt
    case shirt
    case jacket
    case coat
    case dress
    case pants
    case shorts
    case skirt
}

/// エラーコード
public enum ARFitKitError: Error, LocalizedError {
    case initializationFailed
    case cameraAccessDenied
    case gpuNotAvailable
    case modelLoadFailed
    case garmentConversionFailed
    case invalidImage
    case sessionNotStarted
    case networkError
    
    public var errorDescription: String? {
        switch self {
        case .initializationFailed: return "初期化に失敗しました"
        case .cameraAccessDenied: return "カメラへのアクセスが拒否されました"
        case .gpuNotAvailable: return "GPUが利用できません"
        case .modelLoadFailed: return "モデルの読み込みに失敗しました"
        case .garmentConversionFailed: return "衣服の変換に失敗しました"
        case .invalidImage: return "無効な画像です"
        case .sessionNotStarted: return "セッションが開始されていません"
        case .networkError: return "ネットワークエラー"
        }
    }
}

/// セッション設定
public struct SessionConfig {
    public var targetFPS: Int = 60
    public var enableClothSimulation: Bool = true
    public var enableShadows: Bool = true
    public var maxGarments: Int = 3
    public var serverEndpoint: String = ""
    public var useHybridProcessing: Bool = true
    
    public init() {}
}

/// 衣服データ
public class Garment: Identifiable, ObservableObject {
    public let id = UUID()
    public let type: GarmentType
    public let image: UIImage
    
    @Published public var isLoading: Bool = false
    @Published public var conversionProgress: Float = 0.0
    
    internal var nativeGarment: UnsafeMutableRawPointer?
    
    public init(type: GarmentType, image: UIImage) {
        self.type = type
        self.image = image
    }
}

/// ボディポーズデータ
public struct BodyPose {
    public let landmarks: [SIMD3<Float>]
    public let visibility: [Float]
    public let confidence: Float
    public let timestamp: TimeInterval
}

// MARK: - ARFitKit Main Class

/// ARFitKit メインクラス
/// AR試着機能を提供するプライマリインターフェース
public class ARFitKit: NSObject, ObservableObject {
    
    // MARK: - Published Properties
    
    @Published public private(set) var isSessionActive: Bool = false
    @Published public private(set) var currentFPS: Float = 0.0
    @Published public private(set) var currentPose: BodyPose?
    @Published public private(set) var activeGarments: [Garment] = []
    
    // MARK: - Private Properties
    
    private var config: SessionConfig = SessionConfig()
    private var arSession: ARSession?
    private var sceneView: ARSCNView?
    
    private let processingQueue = DispatchQueue(label: "com.arfitkit.processing", qos: .userInteractive)
    private var cancellables = Set<AnyCancellable>()
    
    // Objective-C Bridge
    private let bridge = ARFitKitBridge()
    
    // MARK: - Callbacks
    
    public var onFrameProcessed: ((UIImage) -> Void)?
    public var onPoseUpdated: ((BodyPose) -> Void)?
    public var onError: ((ARFitKitError) -> Void)?
    
    // MARK: - Initialization
    
    public override init() {
        super.init()
        setupCallbacks()
    }
    
    deinit {
        stopSession()
    }
    
    private func setupCallbacks() {
        bridge.onFPSUpdated = { [weak self] fps in
            DispatchQueue.main.async {
                self?.currentFPS = fps
            }
        }
        
        // TODO: Map body tracking bridge callbacks
    }
    
    // MARK: - Public Methods
    
    /// SDKを初期化
    /// - Parameter config: セッション設定
    public func initialize(config: SessionConfig = SessionConfig()) throws {
        self.config = config
        
        let bridgeConfig: [String: Any] = [
            "targetFPS": config.targetFPS,
            "enableClothSimulation": config.enableClothSimulation
        ]
        
        var error: NSError?
        if !bridge.initialize(withConfig: bridgeConfig, error: &error) {
            throw ARFitKitError.initializationFailed
        }
        
        print("ARFitKit initialized with config: \(config)")
    }
    
    /// ARセッションを開始
    /// - Parameter view: AR表示用のビュー
    @MainActor
    public func startSession(view: ARSCNView) async throws {
        guard !isSessionActive else { return }
        
        // Check camera permission
        let status = AVCaptureDevice.authorizationStatus(for: .video)
        switch status {
        case .notDetermined:
            let granted = await AVCaptureDevice.requestAccess(for: .video)
            guard granted else {
                throw ARFitKitError.cameraAccessDenied
            }
        case .denied, .restricted:
            throw ARFitKitError.cameraAccessDenied
        case .authorized:
            break
        @unknown default:
            break
        }
        
        // Setup AR session
        self.sceneView = view
        let configuration = ARBodyTrackingConfiguration()
        configuration.isAutoFocusEnabled = true
        
        view.session.run(configuration)
        view.session.delegate = self
        
        self.arSession = view.session
        self.isSessionActive = true
        bridge.startSession()
        
        print("ARFitKit session started")
    }
    
    /// ARセッションを停止
    public func stopSession() {
        bridge.stopSession()
        arSession?.pause()
        arSession = nil
        sceneView = nil
        isSessionActive = false
        activeGarments.removeAll()
        bridge.removeAllGarments()
        
        print("ARFitKit session stopped")
    }
    
    /// 衣服を読み込む
    /// - Parameters:
    ///   - image: 衣服の画像
    ///   - type: 衣服の種類（省略時は自動検出）
    /// - Returns: 読み込まれた衣服
    public func loadGarment(image: UIImage, type: GarmentType = .unknown) async throws -> Garment {
        return try await withCheckedThrowingContinuation { continuation in
            bridge.loadGarment(image, type: type.rawValue) { [weak self] bridgeGarment, error in
                if let error = error {
                    continuation.resume(throwing: ARFitKitError.modelLoadFailed)
                    return
                }
                
                guard let bridgeGarment = bridgeGarment else {
                    continuation.resume(throwing: ARFitKitError.garmentConversionFailed)
                    return
                }
                
                let garment = Garment(type: type, image: image)
                // In a real app, bind bridgeGarment to garment instance
                
                DispatchQueue.main.async {
                    garment.isLoading = false
                    garment.conversionProgress = 1.0
                    continuation.resume(returning: garment)
                }
            }
        }
    }
    
    /// 衣服を試着
    /// - Parameter garment: 試着する衣服
    public func tryOn(garment: Garment) throws {
        guard isSessionActive else {
            throw ARFitKitError.sessionNotStarted
        }
        
        // Check max garments
        if activeGarments.count >= config.maxGarments {
            activeGarments.removeFirst()
        }
        
        activeGarments.append(garment)
        
        // TODO: Pass garment to bridge
        // bridge.tryOnGarment(garment.bridgeObject)
        
        print("Trying on garment: \(garment.type)")
    }
    
    /// 衣服を削除
    /// - Parameter garment: 削除する衣服
    public func removeGarment(garment: Garment) {
        activeGarments.removeAll { $0.id == garment.id }
        // bridge.removeGarment(...)
    }
    
    /// すべての衣服を削除
    public func removeAllGarments() {
        activeGarments.removeAll()
        bridge.removeAllGarments()
    }
    
    /// スナップショットを撮影
    /// - Returns: 現在のAR画面のスナップショット
    public func captureSnapshot() -> UIImage? {
        return sceneView?.snapshot()
    }
}

// MARK: - ARSessionDelegate

extension ARFitKit: ARSessionDelegate {
    
    public func session(_ session: ARSession, didUpdate frame: ARFrame) {
        // Process frame via Bridge
        bridge.processFrame(frame)
        
        // Note: Real FPS calculation moved to bridge callback
    }
    
    public func session(_ session: ARSession, didUpdate anchors: [ARAnchor]) {
        // Handle body anchor updates
        for anchor in anchors {
            if let bodyAnchor = anchor as? ARBodyAnchor {
                processBodyAnchor(bodyAnchor)
            }
        }
    }
    
    private func processBodyAnchor(_ anchor: ARBodyAnchor) {
        let skeleton = anchor.skeleton
        
        // Convert to BodyPose
        var landmarks: [SIMD3<Float>] = []
        var visibility: [Float] = []
        
        for jointName in skeleton.definition.jointNames {
            if let transform = skeleton.modelTransform(for: ARSkeleton.JointName(rawValue: jointName)) {
                let position = SIMD3<Float>(transform.columns.3.x, 
                                             transform.columns.3.y, 
                                             transform.columns.3.z)
                landmarks.append(position)
                visibility.append(skeleton.isTracked ? 1.0 : 0.0)
            }
        }
        
        let pose = BodyPose(
            landmarks: landmarks,
            visibility: visibility,
            confidence: skeleton.isTracked ? 0.9 : 0.5,
            timestamp: Date().timeIntervalSince1970
        )
        
        DispatchQueue.main.async {
            self.currentPose = pose
            self.onPoseUpdated?(pose)
        }
    }
}

// MARK: - SwiftUI View

import SwiftUI

/// ARFitKit用のSwiftUIビュー
public struct ARFitKitView: UIViewRepresentable {
    @ObservedObject var arFitKit: ARFitKit
    
    public init(arFitKit: ARFitKit) {
        self.arFitKit = arFitKit
    }
    
    public func makeUIView(context: Context) -> ARSCNView {
        let view = ARSCNView()
        view.automaticallyUpdatesLighting = true
        return view
    }
    
    public func updateUIView(_ uiView: ARSCNView, context: Context) {
        // View updates handled by ARFitKit
    }
}
